mlflow/server/js/src/components/CompareRunContour.js

```diff
@@ -0,0 +1,305 @@
+  static MAX_PLOT_KEY_LENGTH = 40;
+  static MAX_PLOT_VALUE_LENGTH = 60;
+
+  constructor(props) {
+    super(props);
+
+    this.entities = new AllHtmlEntities();
+
+    this.metricKeys = CompareRunUtil.getKeys(this.props.metricLists, false);
+    this.paramKeys = CompareRunUtil.getKeys(this.props.paramLists, false);
+
+    if (this.paramKeys.length + this.metricKeys.length < 3) {
+      this.state = {disabled: true};
+    } else {
+      this.state = {
+        disabled: false,
+        reverseColor: false,
+        xaxis: this.paramKeys.length > 0 ?
+        {
+          key: this.paramKeys[0],
+          isMetric: false,
+        } : {
+          key: this.metricKeys[1],
+          isMetric: true,
+        },
+
+        yaxis: this.paramKeys.length > 0 ?
+        {
+          key: this.paramKeys[1],
+          isMetric: false,
+        } : {
+          key: this.metricKeys[1],
+          isMetric: true,
+        },
+
+        zaxis: this.metricKeys.length > 0 ?
+        {
+          key: this.metricKeys[0],
+          isMetric: true,
+        } : {
+          key: this.paramKeys[1],
+          isMetric: false,
+        },
+      };
+    }
+  }
+
+  /**
+   * Get the value of the metric/param described by {key, isMetric}, in run i
+   */
+  getValue(i, {key, isMetric}) {
+    const value = CompareRunUtil.findInList(
+      (isMetric ? this.props.metricLists : this.props.paramLists)[i], key);
+    return value === undefined ? value : value.value;
+  }
+
+  /**
+   * Encode HTML entities in a string (since Plotly's tooltips take HTML)
+   */
+  encodeHtml(str) {
+    return this.entities.encode(str);
+  }
+
+  getColorscale() {
+    /*
+    * contour plot has an option named "reversescale" which
+    * reverses the color mapping if True, but it doesn't work properly now.
+    *
+    * https://github.com/plotly/plotly.js/issues/4430
+    *
+    * This function is a temporary workaround for the issue.
+    */
+    const colorscale = [
+      [0, 'rgb(5,10,172)'],
+      [0.35, 'rgb(40,60,190)'],
+      [0.5, 'rgb(70,100,245)'],
+      [0.6, 'rgb(90,120,245)'],
+      [0.7, 'rgb(106,137,247)'],
+      [1, 'rgb(220,220,220)'],
+    ];
+
+    if (this.state.reverseColor) {
+      return colorscale;
+    } else {
+      // reverse only RGB values
+      return colorscale.map(([val], index) => [
+        val,
+        colorscale[colorscale.length - 1 - index][1],
+      ]);
+    }
+  }
+
+  render() {
+    if (this.state.disabled) {
+      return <div/>;
+    }
+
+    const keyLength = CompareRunContour.MAX_PLOT_KEY_LENGTH;
+
+    const xs = [];
+    const ys = [];
+    const zs = [];
+    const tooltips = [];
+
+    this.props.runInfos.forEach((_, index) => {
+      const x = this.getValue(index, this.state.xaxis);
+      const y = this.getValue(index, this.state.yaxis);
+      const z = this.getValue(index, this.state.zaxis);
+      if (x === undefined || y === undefined || z === undefined) {
+        return;
+      }
+      xs.push(parseFloat(x));
```

Comment:
What happens here if a user tries to plot a param that has non-numeric values? I guess `parseFloat` will return NaN, but how does Plotly handle the result?

I tried to generate & compare a few runs with non-numeric params & view the contour plot but got a JS exception before I was able to try it out. Here's the snippet I used to generate the runs:

```
import mlflow
def create_run():
    with mlflow.start_run():
        mlflow.log_metric("a", 1)
        mlflow.log_param("b", "abc")
        mlflow.log_metric("c", 3)

for i in range(3):
   create_run()
```

And the resulting exception when viewing the Contour plot:

```
index.js:2178 TypeError: Cannot read property 'length' of undefined
    at Function.truncateString (Utils.js:60)
    at CompareRunContour.render (CompareRunContour.js:218)
    at finishClassComponent (react-dom.development.js:8389)
    at updateClassComponent (react-dom.development.js:8357)
    at beginWork (react-dom.development.js:8982)
    at performUnitOfWork (react-dom.development.js:11814)
    at workLoop (react-dom.development.js:11843)
    at renderRoot (react-dom.development.js:11874)
    at performWorkOnRoot (react-dom.development.js:12449)
    at performWork (react-dom.development.js:12370)
    at performSyncWork (react-dom.development.js:12347)
    at interactiveUpdates (react-dom.development.js:12597)
    at interactiveUpdates (react-dom.development.js:1958)
    at dispatchInteractiveEvent (react-dom.development.js:4259)
```

Reply:
The contour plot doesn't work for non-numeric parameters. We need to remove them.